coroutines/src/commonMain/kotlin/DelayStrategy.kt

```diff
@@ -0,0 +1,89 @@
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.transformLatest
+import kotlinx.datetime.Clock
+import kotlinx.datetime.asTimeSource
+import kotlin.math.E
+import kotlin.math.pow
+import kotlin.time.ExperimentalTime
+import kotlin.time.TimeSource
+
+/**
+ * Interface, using the strategy pattern, for an implementer to delay for periods of time.
+ */
+public interface DelayStrategy {
+    public suspend fun await(iteration: Int, elapsedMillis: Long)
+}
+
+/**
+ * DelayStrategy that, when await() is called, will delay for a fixed amount of time, to be determined by [delayMillis].
+ */
+public class FixedDelay(private val delayMillis: Long) : DelayStrategy {
+    override suspend fun await(iteration: Int, elapsedMillis: Long) {
+        delay(delayMillis - elapsedMillis)
+    }
+}
+
+/**
+ * DelayStrategy with an exponentially increasing delay, calculated with [calculateExponentialBackoffDelay].
+ */
+public class ExponentialBackoff(
+    private val minimumMillis: Long,
+    private val maximumMillis: Long
+) : DelayStrategy {
+    override suspend fun await(iteration: Int, elapsedMillis: Long) {
+        delay(calculateExponentialBackoffDelay(iteration, elapsedMillis, minimumMillis, maximumMillis))
+    }
+}
+
+internal fun calculateExponentialBackoffDelay(
+    iteration: Int,
+    elapsedMillis: Long,
+    minimumMillis: Long,
+    maximumMillis: Long
+) = (E.pow(iteration / E).toLong() * 1000L).coerceIn(minimumMillis..maximumMillis) - elapsedMillis
```

Comment:
This exponential function is pretty specific to our needs. As a publicly consumed library, this should be more configurable. Perhaps `EponentialBackoff` can take lambdas for the base and the exponent?

Reply:
While I agree that adding customizability probably makes sense in a public library, it feels awkward to allow for customization on that portion as the delay calculation is the vast majority of the utility of this function.

Maybe the `ExponentialBackoff` class should only exist on the app layer?

Reply:
Alternatively, maybe we could find a more specific name to this particular algorithm?

Reply:
Maybe something along the lines of?:

```kotlin
private const val DEFAULT_MULTIPLIER = 2f
private const val DEFAULT_MAXIMUM_DELAY_MILLIS = Long.MAX_VALUE

/**
 * Exponential backoff implementation using the following formula:
 *
 * ```
 * delay = base * multiplier ^ retry
 * ```
 *
 * For example (using `base = 100` and `multiplier = 2`):
 *
 * | retry | delay |
 * |-------|-------|
 * |   1   |   100 |
 * |   2   |   200 |
 * |   3   |   400 |
 * |   4   |   800 |
 * |   5   |  1600 |
 * |  ...  |   ... |
 *
 * Inspired by:
 * [Exponential Backoff And Jitter](https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/)
 */
class ExponentialBackoff(
    private val baseTimeMillis: Long,
    private val multiplier: Float = DEFAULT_MULTIPLIER,
    maximumDelayMillis: Long = DEFAULT_MAXIMUM_DELAY_MILLIS
) : BackoffStrategy {

    private val maximumMillis = maximumDelayMillis.toFloat()

    override fun getDelayMillis(retry: Int): Long =
        (baseTimeMillis * multiplier.pow(retry - 1))
            .coerceAtMost(maximumMillis)
            .toLong()
}
```

_Pulled from old app code._

Even have some tests to go along with it if you want to pull them in here:

<details>
<summary>Tests<summary>

```kotlin
class ExponentialBackoffTest {

    @Test
    fun defaultsWithMaximumDelayMillis_largeRetryNumber_adheresToMaximum() {
        val backoff = ExponentialBackoff(baseTimeMillis = 100, maximumDelayMillis = 1_000L)
        assertEquals(1_000L, backoff.getDelayMillis(retry = 20))
    }

    @Test
    fun calculatedDelayResultsInInfinity_returnsMaximumDelayMillis() {
        val backoff = ExponentialBackoff(baseTimeMillis = 100_000L, maximumDelayMillis = 1_000L)
        assertEquals(1_000L, backoff.getDelayMillis(retry = 100_000))
    }

    @Test
    fun firstRetry_hasDelayEqualToBaseTimeMillis() {
        val backoff = ExponentialBackoff(baseTimeMillis = 100)
        assertEquals(100, backoff.getDelayMillis(retry = 1))
    }

    @Test
    fun multipleRetries_increaseExponentially() {
        val baseTimeMillis = 100L
        val backoff = ExponentialBackoff(baseTimeMillis)

        var delay = baseTimeMillis
        val retries = 1..10
        retries.forEach { retry ->
            assertEquals(delay, backoff.getDelayMillis(retry), "Backoff retry #$retry")
            delay *= 2
        }
    }
}
```
</details>

Reply:
This is what I was thinking
```kotlin
public class ExponentialBackoff(
    private val minimumMillis: Long,
    private val maximumMillis: Long,
    private val base: (iteration: Int) -> Float,
    private val exponent: (iteration: Int) -> Float,
) : DelayStrategy {
    override suspend fun await(iteration: Int, elapsedMillis: Long) {
        val delay = base.invoke(iteration).pow(exponent.invoke(iteration).coerceIn(minimumMillis..maximumMillis) - elapsedMillis
        delay(delay)
    }
}
```

Reply:
I ended up going with something similar to Travis' proposal. While offering more customization, I found the lambda parameters to largely increase the complexity and required understanding of the public API.

I think this should cover our bases for our Delay implementation